onecodex/models/helpers.py

```diff
@@ -136,25 +138,47 @@ def _download(
         if path and file_obj:
             raise OneCodexException("Please specify only one of: path, file_obj")
 
-        if path is None and file_obj is None:
-            path = os.path.join(os.getcwd(), _filename)
-
-        if path and os.path.exists(path):
-            raise OneCodexException("{} already exists! Will not overwrite.".format(path))
-
         try:
             method_to_call = getattr(self._resource, _resource_method)
+            download_info = method_to_call()
+
+            if path is None and file_obj is None:
+                if _filename is None:
+                    if "filename" not in download_info:
+                        raise OneCodexException(
+                            "Please specify `path`, `file_obj`, or `_filename`."
+                        )
+                    _filename = download_info["filename"]
+                path = os.path.join(os.getcwd(), _filename)
+
+            if path and os.path.exists(path):
+                raise OneCodexException("{} already exists! Will not overwrite.".format(path))
 
             if use_potion_session:
-                resp = self._resource._client.session.get(
-                    method_to_call()["download_uri"], stream=True
-                )
+                session = self._resource._client.session
             else:
-                resp = requests.get(method_to_call()["download_uri"], stream=True)
+                session = requests.Session()
+
+            link = download_info["download_uri"]
+
+            # Retry up to 3 times with backoff timing of 4s, 8s, and 16s (applies to all HTTP
+            # methods). 404 is included for cases where the file is being asynchronously uploaded to
+            # S3 and is expected to be available soon.
```

Comment:
would the 28s be enough for the 404 case?

Reply:
From testing locally it was (I put some `sleep` calls in the task to simulate delays) but it's hard to say if that'll work in the real world (e.g. if the task queue is really overloaded or something). `darwin2-ui` is using 5 retries with a backoff factor of 2. I'll update this code to match `darwin2-ui` since it's been working for ATCC downloads.